custom/sensors/SparkEncoder.java

```diff
@@ -0,0 +1,134 @@
+package org.usfirst.frc4904.standard.custom.sensors;
+
+
+import com.revrobotics.CANEncoder;
+import com.revrobotics.CANSparkMax;
+import org.usfirst.frc4904.standard.Util;
+import edu.wpi.first.wpilibj.PIDSourceType;
+
+public class SparkEncoder implements CustomEncoder {
+    protected final CANEncoder encoder;
+    protected PIDSourceType pidSource;
+    protected double distancePerPulse;
+    protected boolean reverseDirection;
+    protected double truePosition;
+    protected double trueVelocity;
+
+    public SparkEncoder(CANSparkMax spark, boolean reverseDirection, double distancePerPulse) {
+        encoder = new CANEncoder(spark);
+        this.reverseDirection = reverseDirection;
+        this.distancePerPulse = distancePerPulse;
+        setPIDSourceType(PIDSourceType.kDisplacement);
+    }
+
+    @Override
+    public void setPIDSourceType(PIDSourceType pidSource) {
+        this.pidSource = pidSource;
+    }
+
+    @Override
+    public PIDSourceType getPIDSourceType() {
+        return pidSource;
+    }
+
+    @Override
+    public double pidGet() {
+        if (pidSource == PIDSourceType.kDisplacement) {
+            return getDistance();
+        } else {
+            return getRate();
+        }
+    }
+
+    @Override
+    public int get() {
+        return (int) encoder.getPosition();
+    }
+
+    @Override
+    public double getDistance() {
+        if (reverseDirection) {
+            return distancePerPulse * (encoder.getPosition() - truePosition) * -1.0;
+        } else {
+            return distancePerPulse * (encoder.getPosition() - truePosition);
+        }
+    }
+
+    @Override
+    public boolean getDirection() {
+        return !reverseDirection == (encoder.getVelocity() >= 0);
+    }
+
+    @Override
+    public boolean getStopped() {
+        return Util.isZero(getRate());
+    }
+
+    @Override
+    public double getRate() {
```

Comment:
Consider converting from rotations (or distance) per minute to rotations (or distance) per second, or make that a configurable option, since I assume we will usually want it in seconds (except maybe for flywheels?)